meson.build

```diff
@@ -0,0 +1,362 @@
+  ],
+  meson_version : '>=0.49.0',
+)
+
+# TODO: This is redundant with project->version
+flatpak_major_version = 1
+flatpak_minor_version = 13
+flatpak_micro_version = 2
+flatpak_extra_version = ''
+
+flatpak_version = '@0@.@1@.@2@@3@'.format(
+  flatpak_major_version,
+  flatpak_minor_version,
+  flatpak_micro_version,
+  flatpak_extra_version,
+)
+flatpak_interface_age = 0
+flatpak_binary_age = (
+  10000 * flatpak_major_version
+  + 100 * flatpak_minor_version
+  + flatpak_micro_version
+)
+
+required_glib = '2.44'
+required_bwrap = '0.5.0'
+required_dbus_proxy = '0.1.0'
+required_libostree = '2020.8'
+
+gnome = import('gnome')
+i18n = import('i18n')
+pkgconfig = import('pkgconfig')
+
+if meson.version().version_compare('>=0.56')
+  project_build_root = meson.project_build_root()
+  project_source_root = meson.project_source_root()
+else
+  project_build_root = meson.current_build_dir()
+  project_source_root = meson.current_source_dir()
+endif
+
+if meson.version().version_compare('>=0.58')
+  global_source_root = meson.global_source_root()
+else
+  global_source_root = meson.source_root()
+endif
+
+if meson.version().version_compare('>=0.53')
+  fs = import('fs')
+endif
+
+cc = meson.get_compiler('c')
+add_project_arguments('-include', 'config.h', language : 'c')
+common_include_directories = include_directories(
+  '.',
+  'common',
+)
+
+# Keep this in sync with ostree, except remove -Wall (part of Meson
+# warning_level 2) and -Werror=declaration-after-statement
+add_project_arguments(
+  cc.get_supported_arguments([
+    '-Werror=shadow',
+    '-Werror=empty-body',
+    '-Werror=strict-prototypes',
+    '-Werror=missing-prototypes',
+    '-Werror=implicit-function-declaration',
+    '-Werror=pointer-arith',
+    '-Werror=init-self',
+    '-Werror=missing-declarations',
+    '-Werror=return-type',
+    '-Werror=overflow',
+    '-Werror=int-conversion',
+    '-Werror=parenthesis',
+    '-Werror=incompatible-pointer-types',
+    '-Werror=misleading-indentation',
+    '-Werror=missing-include-dirs',
+
+    # Meson warning_level=2 would do this, but we are not fully
+    # signedness-safe yet
+    '-Wno-sign-compare',
+    '-Wno-error=sign-compare',
+
+    # Deliberately not warning about these, ability to zero-initialize
+    # a struct is a feature
+    '-Wno-missing-field-initializers',
+    '-Wno-error=missing-field-initializers',
+
+    # Deliberately not warning about these
+    '-Wno-unused-parameter',
+    '-Wno-error=unused-parameter',
+  ]),
+  language : 'c',
+)
+# Flatpak is Linux-specific, so for now we assume that -fvisibility=hidden
+# is always supported
+add_project_arguments('-fvisibility=hidden', language : 'c')
+
+if (
+  cc.has_argument('-Werror=format=2')
+  and cc.has_argument('-Werror=format-security')
+  and cc.has_argument('-Werror=format-nonliteral')
+)
+  add_project_arguments([
+    '-Werror=format=2',
+    '-Werror=format-security',
+    '-Werror=format-nonliteral',
+  ], language : 'c')
+endif
```

Comment:
What's wrong with get_supported_arguments here too? It saves on duplicating the flag names.

Reply:
To match the Autotools logic, we want to add these three as a group, but only if all three are supported (if only one or two of them are, then we don't want to add any of them).

Reply:
I'm honestly not sure whether these three options need to be clustered together like this... perhaps if we're dropping support for distros older than Debian 10 *anyway*, then we can simplify this. However, I'd prefer to match the Autotools logic for now, because porting the build system to Meson has already required shaving several more yaks than I really wanted to, some of them in the Autotools build...